client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,319 @@
+    }, this.transitionValue || null)().then(() => {
+      // Wait element to transition out and then hide with hidden
+      // eslint-disable-next-line no-param-reassign
+      tabContent.hidden = true;
+    });
+  }
+
+  hideTabContent(tabId: string) {
+    if (tabId === this.selectedValue || !this.selectedValue) {
+      return;
+    }
+
+    const tabContent = this.getTabPanelByHref(tabId);
+    if (!tabContent) {
+      return;
+    }
+    if (this.animateValue) {
+      this.animateOut(tabContent);
+    } else {
+      tabContent.hidden = true;
+    }
+
+    const tab = this.getTabLabelByHref(tabId);
+    if (!tab) {
+      return;
+    }
+    tab.setAttribute('aria-selected', 'false');
+    tab.setAttribute('tabindex', '-1');
+  }
+
+  selectNext(event: Event) {
+    const tabIndex = (event.target as IndexedEventTarget).index;
+    const tab = this.labelTargets[tabIndex + 1];
+    if (!tab) {
+      return;
+    }
+    this.selectedValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  selectPrevious(event: Event) {
+    const tabIndex = (event.target as IndexedEventTarget).index;
+    const tab = this.labelTargets[tabIndex + -1];
+    if (!tab) {
+      return;
+    }
+    this.selectedValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  selectFirst() {
+    const tab = this.labelTargets[0];
+    this.selectedValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  selectLast() {
+    const tab = this.labelTargets[this.labelTargets.length - 1];
+    this.selectedValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  loadHistory(event: PopStateEvent) {
+    if (event.state && event.state.tabContent) {
+      const tab = this.getTabLabelByHref(event.state.tabContent);
+      if (tab) {
+        this.selectedValue = event.state.tabContent;
+        tab.focus();
+      }
+    }
+  }
+
+  validate() {
+    this.labelTargets.forEach((label, idx) => {
+      const panel = this.panelTargets[idx];
+      if (label.getAttribute('role') !== 'tab') {
+        // eslint-disable-next-line no-console
+        console.warn(
+          label,
+          "this element does not have role='tab' aria attribute",
+        );
+      }
+      if (panel.getAttribute('role') !== 'tabpanel') {
+        // eslint-disable-next-line no-console
+        console.warn(
+          panel,
+          "this element does not have role='tabpanel' aria attribute",
+        );
+      }
+      if (panel.getAttribute('aria-labelledby') !== label.id) {
+        // eslint-disable-next-line no-console
+        console.warn(panel, 'this element does not have aria-labelledby');
+      }
+      if (this.listTarget.getAttribute('role') !== 'tablist') {
+        // eslint-disable-next-line no-console
+        console.warn(
+          this.listTarget,
+          "this element does not have role='tablist' aria attribute",
+        );
+      }
+    });
```

Comment:
I think we can remove the need for the listTarget, this means less Stimulus markup needed and we can easily check all 'label' elements have the correct parent.


```suggestion
      if (
        this.labelTargets.every(
          (target) =>
            (target.parentElement as HTMLElement).getAttribute('role') !==
            'tablist',
        )
      ) {
        // eslint-disable-next-line no-console
        console.warn(
          this.labelTargets,
          "One or more tab (label) targets are not direct descendants of an element with role='tablist'.",
        );
      }
```